lib/client.js

```diff
@@ -941,6 +950,12 @@ function onSocketError (err) {
 }
 
 function onSocketEnd () {
+  // @todo move to util function?
+  if (this[kClient][kBodySize] > 0 && this[kClient][kResponseContentLength] > 0 &&
+    this[kClient][kBodySize] !== this[kClient][kResponseContentLength]) {
+    util.destroy(this, new ResponseContentLengthMismatchError('end'))
+    return
+  }
```

Comment:
This check should be in onBody and onMessageComplete

Reply:
Should probably also check for the strictcontentlength flag

Reply:
That flag refers to the sending headers, I think we should rename it, because in the current situation there are 2 cases and both return an error:

- the server content-length is greater than the actual body => connection is closed by the server, client returns "UND_ERR_SOCKET" >> fix sending "UND_ERR_RESPONSE_CONTENT_LENGTH_MISMATCH"
- the server content-length is less than the actual body => connection is closed by the client, client returns "UND_ERR_INFO" >> ? fix sending "UND_ERR_RESPONSE_CONTENT_LENGTH_MISMATCH" 

We can add another flag to suppress these errors, because I think those flags have different meanings: as client we can control what we send, so if we see an error/warning on sending headers we can fix it, but we can't control what the server is sending.

I think the flags should be something like

```
requestStrictContentLength: 'warn' (or 'error' or 'silent')
responseStrictContentLength: 'warn' (or 'error' or 'silent')
```

Reply:
The only place the check works is in `onSocketEnd`, doing in `onMessage` and/or `onBody` don't work, and also break other tests.

I think we can't use the `strictContentLength` option, or another one, because the connection always ends with an error, so we can only send the right one instead of the generic

Reply:
Doing it in onSocketEnd is unfortunately just plain wrong. It should basically only be call if the server unexpectedly closes the connection.

Reply:
I don't see other solutions

Reply:
> doing in onMessage and/or onBody don't work, and also break other tests.

In what way doesn't it work?

Reply:
> and onSocketEnd is actually called when the server close the connection unexpectedly

Yes, and what if the server doesn't close the connection when there is a content length mismatch?

Reply:
> > doing in onMessage and/or onBody don't work, and also break other tests.
> 
> In what way doesn't it work?

It would take me some time to answer this, like re-implementing in that way. The main issue is that `onMessage` and `onBody` are not been called when the connection is closed

Reply:
> > and onSocketEnd is actually called when the server close the connection unexpectedly
> 
> Yes, and what if the server doesn't close the connection when there is a content length mismatch?

Sorry, I don't get it. As far as I've seen, the `content-length mismatch` is caused by the server when it close the connection before sending the "promised" body

Reply:
Ah, now I see where the confusion is. I’m talking about the case where a misbehaving server sends a larger body than expected.

Reply:
Or if the server sends a shorter body without disconnecting

Reply:
The disconnecting is a different error

Reply:
I've seen body can't be larger than content-length because of https://github.com/nodejs/undici/blob/e2a9ddc07d9e0ff6be25d243523b53d1d71a3e9e/deps/llhttp/include/llhttp.h - unless we want to add this option

if the server send a shorter body without disconnecting, the timeout error is sent

is that right? Am I missing something?

Reply:
With keep-alive the server can give a body that doesn't match content-length.

Reply:
> I've seen body can't be larger than content-length because of https://github.com/nodejs/undici/blob/e2a9ddc07d9e0ff6be25d243523b53d1d71a3e9e/deps/llhttp/include/llhttp.h - unless we want to add this option

I think you were intending to link to a specific line?

Reply:
yes, sorry https://github.com/nodejs/undici/blob/e2a9ddc07d9e0ff6be25d243523b53d1d71a3e9e/deps/llhttp/include/llhttp.h#L472

Reply:
> With keep-alive the server can give a body that doesn't match content-length without closing the connection.

ok, I'll work on that

Reply:
@ronag to me it's working fine also with keep-alive option  

https://github.com/nodejs/undici/pull/742/files#diff-3058feaff6f3b2902370ee08a7528340aa7a98e5b76063f0176cb61972fadbb4R347